queries/arithmetic.py

```diff
@@ -0,0 +1,410 @@
+
+
+# Map operator name to python operator
+_STD_OPERATORS = {"multiply": "*", "divide": "/", "plus": "+", "minus": "-"}
+
+# Number of args required for each operator
+_OP_NUM_ARGS = {
+    "multiply": 2,
+    "divide": 2,
+    "plus": 2,
+    "minus": 2,
+    "sqrt": 1,
+    "pow": 2,
+    "percent": 2,
+}
+
+
+def calc_arithmetic(query, result):
+    """ Calculate the answer to an arithmetic query """
+    operator = result.operator
+    nums = result.numbers
+    desc = result.desc
+
+    # Ensure that we have the right number of
+    # number args for the operation in question
+    assert _OP_NUM_ARGS[operator] == len(nums)
+
+    # Global namespace for eval
+    # Block access to all builtins
+    eval_globals = {"__builtins__": None}
+
+    def err_answer(errstr):
+        return dict(answer=errstr), errstr, errstr
+
+    # Square root calculation
+    if operator == "sqrt":
+        if len(str(nums[0])) > 100:
+            return err_answer("Þessi tala er of há.")
+        # Allow sqrt function in eval namespace
+        eval_globals["sqrt"] = math.sqrt
+        s = "sqrt({0})".format(nums[0])
+
+    # Pow
+    elif operator == "pow":
+        # Cap max pow
+        if nums[1] > 50:
+            return err_answer("Þetta er of hátt veldi.")
+        # Allow pow function in eval namespace
+        eval_globals["pow"] = pow
+        s = "pow({0},{1})".format(nums[0], nums[1])
+
+    # Percent
+    elif operator == "percent":
+        s = "({0} * {1}) / 100.0".format(nums[0], nums[1])
+
+    # Addition, subtraction, multiplication, division
+    elif operator in _STD_OPERATORS:
+        math_op = _STD_OPERATORS[operator]
+
+        # Check for division by zero
+        if math_op == "/" and nums[1] == 0:
+            return err_answer("Það er ekki hægt að deila með núlli.")
+
+        s = "{0} {1} {2}".format(nums[0], math_op, nums[1])
+    else:
+        logging.warning("Unknown operator: {0}".format(operator))
+        return None
+
+    # Set arithmetic expression as query key
+    result.qkey = s
+
+    # Run eval on expression
+    res = eval(s, eval_globals, {})
+
+    if isinstance(res, float):
+        # Convert result to Icelandic decimal format
+        answer = format_icelandic_float(res)
+    else:
+        answer = str(res)
+
+    response = dict(answer=answer)
+    voice_answer = "{0} er {1}".format(desc, answer)
+
+    return response, answer, voice_answer
+
+
+def sentence(state, result):
+    """ Called when sentence processing is complete """
+    q = state["query"]
+    if "qtype" in result:
+        # Successfully matched a query type
+        q.set_qtype(result.qtype)
+
+        try:
+            r = calc_arithmetic(q, result)
+            if r is not None:
+                q.set_answer(*r)
+                q.set_key(result.get("qkey"))
+            else:
+                raise Exception("Arithmetic calculation failed")
+        except AssertionError:
+            raise
+        except Exception as e:
+            raise
```

Comment:
Need to choose whether to re-raise or to do `q.set_error()` (the latter is probably correct for release)